### PR TITLE
Avoid stack overflow in Azure NextGen program examples

### PR DESCRIPTION
**Pull request by mikhailshilkov**
Extends https://github.com/pulumi/pulumi/pull/5404 to fix https://github.com/pulumi/pulumi/issues/5228 for all Azure NextGen program examples

### CODE REVIEW

File: pkg/codegen/hcl2/model/type_test.go [hunk 1]
```diff
@@ -668,20 +668,32 @@ func TestUnifyType(t *testing.T) {
 }
 
 func TestRecursiveObjectType(t *testing.T) {
-	props := map[string]Type{
-		"data": NewOutputType(IntType),
+	makeType := func(prop string) Type {
```
Comment (viveklak) [inline pkg/codegen/hcl2/model/type_test.go:671]: Nit - rename to `makeLinkedListType`?

File: pkg/codegen/hcl2/model/type_object.go [hunk 2]
```diff
@@ -218,20 +227,35 @@ func (t *ObjectType) conversionFrom(src Type, unifying bool) ConversionKind {
...
+	if t.s != "" {
+		return t.s
+	}
 
-		annotations := ""
-		if len(t.Annotations) != 0 {
-			annotations = fmt.Sprintf(", annotated(%p)", t)
+	if seen != nil {
+		if _, ok := seen[t]; ok {
+			return "..."
```
Comment (viveklak) [inline pkg/codegen/hcl2/model/type_object.go:240]: Can it be the case that t.s is not set but the object type exists in `seen`? Is that a panic worthy event? What does `...` actually mean in hcl2?

File: pkg/codegen/hcl2/model/type_test.go (hunk 1 above)
Comment (viveklak) [reply pkg/codegen/hcl2/model/type_test.go:671]: Should we consider adding some more tests for unions etc.?

**Review by viveklak — COMMENTED** [review]
Generally looks good apart from a couple of things pointed out in comments.

File: pkg/codegen/hcl2/model/type_object.go (hunk 2 above)
Comment (mikhailshilkov) [reply pkg/codegen/hcl2/model/type_object.go:240]: > Can it be the case that t.s is not set but the object type exists in seen?

Yes - that's when the same type is the "child" of itself - as seen in the test https://github.com/pulumi/pulumi/pull/6250/files#diff-eff2ea294a07fa6b0e4ca9b4e2766d60c1e923cad0771d3deb9514d9b3a4036dR693

> What does ... actually mean in hcl2?

It means nothing in HCL. We aren't producing HCL here, we are producing a text representation of a type.

File: pkg/codegen/hcl2/model/type_test.go (hunk 1 above)
Comment (mikhailshilkov) [reply pkg/codegen/hcl2/model/type_test.go:671]: We could... I mostly extended the existing set of recursive tests to the new function. I don't think it makes the coverage lower but it's true it doesn't improve it either.

**Review by viveklak — APPROVED** [review]

**Pull request merged**